File: coding-task/src/test/scala/ImmutableQueueTest.scala
import org.scalatest.FunSuite

class ImmutableQueueTest extends FunSuite {

  test("element should be empty when one element queue is dequeued") {
    val intQueue = ImmutableQueue[Int](1)
    assert(intQueue.deQueue().isEmpty)
  }

  test("enQueue method should add an element to the end of the queue") {
    val intQueue = ImmutableQueue[Int](1, 2)
    assert(intQueue.enQueue(3) == ImmutableQueue[Int](1, 2, 3))
  }

  test("enQueue method should return new instance") {
    val intQueue1 = ImmutableQueue[Int](1, 2)
    val intQueue2 = ImmutableQueue[Int](1, 2, 3)
    assert(intQueue1.enQueue(3) ne intQueue2)
  }

  test("deQueue method should return queue excluding the first element") {
    val intQueue1 = ImmutableQueue[Int](1, 2, 3)
    assert(intQueue1.deQueue() == ImmutableQueue[Int](2, 3))
  }

  test("deQueue method should return new instance") {
    val intQueue1 = ImmutableQueue[Int](1, 2, 3)
    val intQueue2 = ImmutableQueue[Int](2, 3)
    assert(intQueue1.deQueue() ne intQueue2)
  }

  test("head method should return first element") {
    val intQueue = ImmutableQueue[Int](1, 2)
    assert(intQueue.head === Some(1))

    val stringQueue = ImmutableQueue[String]("foo", "bar")
    assert(stringQueue.head === Some("foo"))
  }
}

File: coding-task/java/src/test/java/ImmutableQueueTest.java
import org.junit.Test;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.NoSuchElementException;

import static org.junit.Assert.*;

public class ImmutableQueueTest {

    private Queue<Integer> emptyQueue = new ImmutableQueue<>(new ArrayList<>());

    @Test
    public void testIsEmpty() {
        assertTrue(emptyQueue.isEmpty());

        Queue<Integer> integerQueue = new ImmutableQueue<>(Arrays.asList(1, 2));
        assertFalse(integerQueue.isEmpty());
    }

    @Test
    public void testEnqueue() {
        Queue<Integer> integerQueue = emptyQueue.enQueue(1);
        assertEquals(integerQueue, new ImmutableQueue<>(Arrays.asList(1)));
        assertNotSame(integerQueue, new ImmutableQueue<>(Arrays.asList(1)));
    }

    @Test(expected = NoSuchElementException.class)
    public void testDequeueToEmptyQueue() {
        emptyQueue.deQueue();
    }

    @Test
    public void testDequeue() {
        Queue<Integer> integerQueue = new ImmutableQueue<>(Arrays.asList(1, 2, 3));
        assertEquals(integerQueue.deQueue(), new ImmutableQueue<>(Arrays.asList(2, 3)));
        assertNotSame(integerQueue.deQueue(), new ImmutableQueue<>(Arrays.asList(2, 3)));
    }

    @Test
    public void testHead() {
        assertNull(emptyQueue.head());
        Queue<Integer> integerQueue = new ImmutableQueue<>(Arrays.asList(1, 2, 3));
        assertEquals(integerQueue.head(), Integer.valueOf(1));
    }

}
